jax/_src/pjit.py

```diff
@@ -887,9 +888,139 @@ def _process_in_axis_resources(in_shardings_thunk, in_layouts_thunk, in_avals,
+      p("  closest seen is passing no keyword args")
+    else:
+      p(f"  closest seen passes {len(close_kwargs)} keyword args with keys:\n"
+        f"    {', '.join(map(repr, close_kwargs))}")
+    return done()
+
+  # have we never seen this tracing context before?
+  ctxs_match = [c for *_, c in seen_keys if c == ctx]
+  if not ctxs_match:
+    p("  tracing context doesn't match, e.g. due to config or context manager")
+    dont_match = [c for *_, c in seen_keys if c != ctx]
+    closest_ctx = min(dont_match, key=lambda c: sum(map(op.ne, c, ctx)))
+    idxs = [i for i, (c1, c2) in enumerate(zip(ctx, closest_ctx)) if c1 != c2]
+    p("  closest seen context tuple differs at positions:\n"
+      f"    {', '.join(map(str, idxs))}\n"
+      "  compare to tuple returned by config._trace_context() in jax/_src/config.py.")
+    return done()
+
+  # have we never seen this input pytree before?
+  trees_match = [k for k in seen_keys if k[0] == in_tree]
+  if not trees_match:
+    in_tree_str = f':\n    {in_tree}' if len(str(in_tree)) < 76 else ''
+    p(f"  never seen input pytree{in_tree_str}")
+    dont_match = [t for t, *_ in seen_keys if t != in_tree]
+    closest_tree = min(dont_match, key=lambda t: abs(t.num_leaves - in_tree.num_leaves))
+    # TODO(mattjj): make equality_errors not print type name, avoid metaclass
+    leaf = type('LeafMeta', (type,), dict(__repr__=lambda _: 'leaf'))('Leaf', (), {})()
+    this_dummy = tree_unflatten(in_tree, [leaf] * in_tree.num_leaves)
+    close_dummy = tree_unflatten(closest_tree, [leaf] * closest_tree.num_leaves)  # type: ignore
+    errs = list(tree_util.equality_errors(this_dummy, close_dummy))
+    p(f"  closest seen input pytree has {len(errs)} mismatches, including:")
+    for path, thing1, thing2, explanation in errs:
+      fst, *path = path  # type: ignore
+      base = ['args', 'kwargs'][fst.idx]
+      p(f"    * at {base}{keystr(path)}, seen {thing2} but now given {thing1},"
+        f"      so {explanation}")
+    return done()
+
+  # have we never seen these input types (eg shapes, dtypes) before?
+  types_match = [k for k in trees_match if k[1] == in_type]
+  if not types_match:
+    if len(in_type) < 5:
+      in_type_str = ':\n    {}'.format(',  '.join(
+          f'{n}: {ty.str_short(short_dtypes=True)}'
+          for n, ty in zip(debug_info.arg_names, in_type)))
+    else:
+      in_type_str = ''
+    p(f"  never seen input type signature{in_type_str}")
+    dont_match = [t for _, t, *_ in trees_match if t != in_type]
+    closest_ty = min(dont_match, key=lambda t: sum(map(op.ne, t, in_type)))
+    num_mismatch = sum(map(op.ne, closest_ty, in_type))
+    p(f"  closest seen input type signature has {num_mismatch} mismatches, including:")
+    add_weak_type_hint = False
+    for name, ty1, ty2 in zip(debug_info.arg_names, closest_ty, in_type):
+      if ty1 != ty2:
+        if type(ty1) == type(ty2) == core.ShapedArray:
+          s1, s2 = ty1.str_short(True), ty2.str_short(True)
```

Comment:
Doesn't str(ShapedArray) already print weak_type? So why do we need this extra logic around weak_type?

Reply:
It's because `str(ShapedArray)` is long and contains info that we shouldn't actually expose to users (ie like "ShapedArray" itself), e.g. `str(ShapedArray((2, 3), 'float32')) == "ShapedArray(float32[2,3])"`. Printing to the user is what `AbstractValue.str_short()` is for, e.g. `ShapedArray((2, 3), 'float32').str_short(True) == "f32[2,3]", just the facts the user wants to see!

Reply:
Oh also, I think we shouldn't print weak_type information unless it's actually relevant. It's just clutter if there are no weak types changing.

Reply:
Then shouldn't we use str_short() everywhere? I see you are using `str` in some places.